src/main/scala/beam/analysis/cartraveltime/CarRideStatsFromPathTraversalEventHandler.scala

```diff
@@ -0,0 +1,301 @@
+    writeIterationCarRideStats(event, carRideStatistics)
+
+    carPtes.clear()
+  }
+
+  private def writeRideStats(iterationNumber: Int, rideStats: Seq[SingleRideStat]): Unit = {
+    val maybeOutputPath = maybeControlerIO.map(cio => cio.getIterationFilename(iterationNumber, "CarRideStats.csv.gz"))
+    maybeOutputPath.foreach { outputPath =>
+      val csvWriter =
+        new CsvWriter(outputPath, Vector("vehicle_id", "travel_time", "distance", "free_flow_travel_time"))
+      try {
+        rideStats.foreach { stat =>
+          csvWriter.write(stat.vehicleId, stat.travelTime, stat.distance, stat.freeFlowTravelTime)
+        }
+      } catch {
+        case NonFatal(ex) =>
+          logger.error(s"Writing ride stats to the ${outputPath} has failed with: ${ex.getMessage}", ex)
+      } finally {
+        Try(csvWriter.close())
+      }
+    }
+  }
+
+  override def notifyShutdown(event: ShutdownEvent): Unit = {
+    toClose.foreach(c => Try(c.close()))
+  }
+
+  private def writeIterationCarRideStats(event: IterationEndsEvent, carRideStatistics: IterationCarRideStats): Unit = {
+    // Write car travel time stats to CSV
+    maybeTravelTimeStatsWriter.foreach(writeStats(_, event.getIteration, carRideStatistics.travelTime.stats))
+    // Write car travel distance stats to CSV
+    maybeTravelDistanceStatsWriter.foreach(writeStats(_, event.getIteration, carRideStatistics.distance.stats))
+    // Write car travel speed stats to CSV
+    maybeTravelSpeedStatsWriter.foreach(writeStats(_, event.getIteration, carRideStatistics.speed.stats))
+    // Write free flow car travel time stats to CSV
+    maybeFreeFlowTravelTimeStatsWriter.foreach(
+      writeStats(_, event.getIteration, carRideStatistics.freeFlowTravelTime.stats)
+    )
+    // Write free flow car speed stats to CSV
+    maybeFreeFlowTravelSpeedStatsWriter.foreach(
+      writeStats(_, event.getIteration, carRideStatistics.freeFlowSpeed.stats)
+    )
+  }
+
+  private def writeStats(csvWriter: CsvWriter, iteration: Int, statistics: Statistics): Unit = {
+    try {
+      csvWriter.write(
+        iteration,
+        statistics.avg,
+        statistics.median,
+        statistics.p75,
+        statistics.p95,
+        statistics.p99,
+        statistics.minValue,
+        statistics.maxValue,
+        statistics.sum
+      )
+      csvWriter.flush()
+    } catch {
+      case NonFatal(ex) =>
+        logger.error(s"Could not write iteration $iteration stats ${statistics}. Error: ${ex.getMessage}", ex)
+    }
+  }
+}
+
+object CarRideStatsFromPathTraversalEventHandler extends LazyLogging {
+
+  def eventsFilterWhenReadFromCsv(event: Event): Boolean = {
+    // We need only PathTraversal
+    val isNeededEvent = event.getEventType == "PathTraversal"
+    isNeededEvent
+  }
+
+  def apply(pathToNetwork: String, eventsFilePath: String): CarRideStatsFromPathTraversalEventHandler = {
+    val network: Network = {
+      val n = NetworkUtils.createNetwork()
+      new MatsimNetworkReader(n)
```

Comment:
Does this read the path into the network (i.e. mutable?) Otherwise what's the point of this line?

Reply:
`MatsimNetworkReader.readFile()` returns `void`, but it mutates the passed network. That's the reason